src/main/java/net/mcreator/preferences/PreferencesManager.java

```diff
@@ -20,42 +21,204 @@
+import net.mcreator.plugin.events.ApplicationLoadedEvent;
+import net.mcreator.preferences.data.Preferences;
+import net.mcreator.preferences.entries.PreferenceEntry;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import java.awt.*;
 import java.io.File;
+import java.util.List;
+import java.util.*;
 
 public class PreferencesManager {
 
 	private static final Logger LOG = LogManager.getLogger("Preferences Manager");
 
-	public static PreferencesData PREFERENCES = new PreferencesData();
+	private static final File file = UserFolderManager.getFileFromUserFolder("preferences.json");
 
 	private static final Gson gson = new GsonBuilder().setPrettyPrinting().setLenient().create();
-	private static final File preferencesFile = UserFolderManager.getFileFromUserFolder("preferences");
 
-	public static void loadPreferences() {
-		if (!UserFolderManager.getFileFromUserFolder("preferences").isFile()) {
-			storePreferences(new PreferencesData());
-			LOG.info("Preferences not created yet. Loading defaults.");
+	/**
+	 * <p>Store values when the preferences' file is loaded</p>
+	 */
+	private static Map<String, Map<String, List<LinkedTreeMap<?, ?>>>> loadedPreferences;
+
+	/**
+	 * <p>Store all preferences using an identifier </p>
+	 */
+	private static Map<String, List<PreferenceEntry<?>>> PREFERENCES_REGISTRY;
+
+	/**
+	 * <p>Default values for all preferences. Used when resetting preferences.</p>
+	 */
+	private static Map<String, List<PreferenceEntry<?>>> DEFAULT_PREFERENCES;
+
+	/**
+	 * <p>MCreator's preferences</p>
+	 */
+	public static Preferences PREFERENCES;
+
+	/**
+	 * <p>Init the system and load MCreator's preferences, so entries used inside the launcher can be get.</p>
+	 */
+	public static void init() {
+		PREFERENCES_REGISTRY = new HashMap<>();
+		DEFAULT_PREFERENCES = new HashMap<>();
+		PREFERENCES = new Preferences();
+
+		if (!file.isFile() && UserFolderManager.getFileFromUserFolder("preferences").exists()) {
+			LOG.debug("Old preferences detected. Converting them to the new format.");
+			convertOldPreferences();
 		} else {
-			try {
-				PREFERENCES = gson.fromJson(FileIO.readFileToString(preferencesFile), PreferencesData.class);
-				if (PREFERENCES == null)
-					throw new NullPointerException("Preferences are null!");
-				LOG.debug("Loading preferences from " + preferencesFile);
-			} catch (Exception e) {
-				LOG.error("Failed to load preferences. Reloading defaults!", e);
-				storePreferences(new PreferencesData());
+			loadedPreferences = gson.fromJson(FileIO.readFileToString(file), Map.class);
+			loadPreferences("mcreator");
+		}
+	}
+
+	/**
+	 * <p>Once plugins are loaded, we can now load preferences registered by them with {@link ApplicationLoadedEvent}.</p>
+	 */
+	public static void loadPlugins() {
+		PREFERENCES_REGISTRY.forEach((identifier, preferences) -> {
+			if (!identifier.equals("mcreator"))
+				loadPreferences(identifier);
+		});
+		savePreferences();

```

Comment:
Is this save here needed? If so, maybe add a comment why

Reply:
This was mostly in case there is an error and an exception is caught, when loading an identifier's preferences. This way it would not be logged each time MCreator is launched and values are not constantly reset.

Reply:
Maybe in this case loadPreferences should throw some exception and it can be caught here and then you save in catch if needed

Reply:
I did some more tests and I noticed preferences are automatically saved when closing MCreator, so this is not really needed. However, I added a try catch when reading the file as MCreator crashed when the JSOn file had a syntax error.